cpp/src/arrow/filesystem/filesystem.cc

```diff
@@ -738,6 +761,109 @@ Result<std::shared_ptr<FileSystem>> FileSystemFromUriReal(const Uri& uri,
 
 }  // namespace
 
+Status RegisterFileSystemFactory(std::vector<std::string> schemes,
+                                 FileSystem::Factory factory) {
+  auto& [mutex, scheme_to_factory] = FileSystemFactoryRegistry();
+  std::unique_lock lock{mutex};
+
+  for (auto&& scheme : schemes) {
+    if (auto& ref = scheme_to_factory[scheme]) {
+      return Status::KeyError(
+          "Tried to add a factory for ", scheme,
+          ":// uris, but a factory was already registered for that scheme");
+    } else {
+      ref = factory;
+    }
+  }
+  return Status::OK();
+}
+
+// XXX alternative approach:
+// If we distrust global constructors more than we distrust
+// symbol visibility, we could export specially named functions
+// and retrieve them via dlsym.
+Status RegisterFileSystemFactoryModule(std::string module_name) {
```

Comment:
I'm not sure why we would want to expose something like this. Why not let third parties choose how they want to load their own code?

Reply:
FTR: The registrar approach is my preference rather than this. However extern "C" module functions are a common pattern ([DuckDB](https://github.com/duckdb/arrow/blob/main/src/arrow_extension.cpp#L34-L43), [NodeJS](https://nodejs.org/api/addons.html#context-aware-addons), [Ruby](https://guides.rubygems.org/gems-with-extensions/), ...) and I included them for completeness.

Reply:
Third parties can freely use `RegisterFileSystemFactory` from their DLL initialization functions or static members. Why would we need to expose something specific here?